File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/http/GraphAnalyzer.kt
/** Graph analyzer — server-side graph structure analysis for KB. */
package com.codeintel.http

import kotlinx.serialization.Serializable
import java.sql.Connection
import java.time.Instant
import java.time.temporal.ChronoUnit

class GraphAnalyzer(private val conn: Connection) {

    fun analyze(): GraphAnalysisResult {
        val stats = computeStats()
        val insights = mutableListOf<GraphInsight>()
        insights.addAll(findOrphans())
        insights.addAll(findHubs())
        insights.addAll(findClusters())
        insights.addAll(findStaleNodes())
        return GraphAnalysisResult(insights, stats, Instant.now().toString())
    }

    private fun computeStats(): GraphStats {
        return runCatching {
            val nc = queryInt("SELECT COUNT(*) FROM knowledge_entries")
            val ec = queryInt("SELECT COUNT(*) FROM knowledge_graph_edges")
            val maxEdges = if (nc > 1) nc.toLong() * (nc - 1) else 1L
            val density = (ec.toDouble() / maxEdges * 10000).toLong() / 10000.0
            GraphStats(nc, ec, density)
        }.getOrDefault(GraphStats(0, 0, 0.0))
    }

    private fun findOrphans(): List<GraphInsight> {
        return runCatching {
            val rs = conn.createStatement().executeQuery(
                "SELECT e.id FROM knowledge_entries e WHERE e.id NOT IN (" +
                "  SELECT source_id FROM knowledge_graph_edges " +
                "  UNION SELECT target_id FROM knowledge_graph_edges) LIMIT 20"
            )
            val ids = buildList { while (rs.next()) add(rs.getLong("id")) }
            if (ids.isEmpty()) return emptyList()
            listOf(GraphInsight(
                type = "orphans", title = "${ids.size} Orphan Nodes",
                description = "Entries không có relationships nào",
                nodeIds = ids, severity = "warning",
                action = InsightAction("Find Related", "api/kb/entries/{id}/find-related", "POST"),
            ))
        }.getOrDefault(emptyList())
    }

    private fun findHubs(): List<GraphInsight> {
        return runCatching {
            val rs = conn.createStatement().executeQuery(
                "SELECT node_id, SUM(cnt) as total FROM (" +
                "  SELECT source_id AS node_id, COUNT(*) AS cnt FROM knowledge_graph_edges GROUP BY source_id" +
                "  UNION ALL" +
                "  SELECT target_id AS node_id, COUNT(*) AS cnt FROM knowledge_graph_edges GROUP BY target_id" +
                ") GROUP BY node_id HAVING total > 10 ORDER BY total DESC LIMIT 10"
            )
            val ids = buildList { while (rs.next()) add(rs.getLong("node_id")) }
            if (ids.isEmpty()) return emptyList()
            listOf(GraphInsight(
                type = "hubs", title = "${ids.size} Hub Nodes",
                description = "Entries có >10 relationships (highly connected)",
                nodeIds = ids, severity = "info", action = null,
            ))
        }.getOrDefault(emptyList())
    }

    private fun findClusters(): List<GraphInsight> {
        return runCatching {
            val edges = queryEdges()
            val nodeIds = queryNodeIds()
            if (nodeIds.isEmpty()) return emptyList()
            val adj = nodeIds.associateWith { mutableSetOf<Long>() }.toMutableMap()
            for ((src, tgt) in edges) {
                adj[src]?.add(tgt)
                adj[tgt]?.add(src)
            }
            val components = countComponents(adj)
            if (components <= 1) return emptyList()
            listOf(GraphInsight(
                type = "clusters", title = "$components Disconnected Clusters",
                description = "Graph có nhiều components tách biệt",
                nodeIds = emptyList(), severity = "info", action = null,
            ))
        }.getOrDefault(emptyList())
    }

    private fun findStaleNodes(): List<GraphInsight> {
        val threshold = Instant.now().minus(180, ChronoUnit.DAYS).toString()
        return runCatching {
            val stmt = conn.prepareStatement(
                "SELECT id FROM knowledge_entries WHERE updated_at < ? LIMIT 15"
            )
            stmt.setString(1, threshold)
            val rs = stmt.executeQuery()
            val ids = buildList { while (rs.next()) add(rs.getLong("id")) }
            if (ids.isEmpty()) return emptyList()
            listOf(GraphInsight(
                type = "stale", title = "${ids.size} Stale Nodes (>180 days)",
                description = "Entries chưa được update > 180 ngày",
                nodeIds = ids, severity = "warning",
                action = InsightAction("Review", "api/kb/entries/{id}/review", "POST"),
            ))
        }.getOrDefault(emptyList())
    }

    private fun queryInt(sql: String): Int {
        val rs = conn.createStatement().executeQuery(sql)
        rs.next()
        return rs.getInt(1)
    }

    private fun queryEdges(): List<Pair<Long, Long>> {
        val rs = conn.createStatement().executeQuery("SELECT source_id, target_id FROM knowledge_graph_edges")
        return buildList { while (rs.next()) add(rs.getLong("source_id") to rs.getLong("target_id")) }
    }

    private fun queryNodeIds(): Set<Long> {
        val rs = conn.createStatement().executeQuery("SELECT id FROM knowledge_entries")
        return buildSet { while (rs.next()) add(rs.getLong("id")) }
    }
}

/** Count connected components via BFS. */
private fun countComponents(adj: Map<Long, Set<Long>>): Int {
    val visited = mutableSetOf<Long>()
    var components = 0
    for (node in adj.keys) {
        if (node in visited) continue
        components++
        val queue = ArrayDeque<Long>()
        queue.add(node)
        while (queue.isNotEmpty()) {
            val current = queue.removeFirst()
            if (!visited.add(current)) continue
            adj[current]?.forEach { if (it !in visited) queue.add(it) }
        }
    }
    return components
}

@Serializable
data class GraphInsight(
    val type: String,
    val title: String,
    val description: String,
    val nodeIds: List<Long>,
    val severity: String,
    val action: InsightAction?,
)

@Serializable
data class InsightAction(val label: String, val endpoint: String, val method: String)

@Serializable
data class GraphStats(val nodeCount: Int, val edgeCount: Int, val density: Double)

@Serializable
data class GraphAnalysisResult(
    val insights: List<GraphInsight>,
    val stats: GraphStats,
    val computedAt: String,
)

File: backend/mcp-code-intelligence-nodejs/src/http/graph-analyzer.ts
/**
 * GraphAnalyzer — server-side graph structure analysis for KB.
 * Port of Python graph_analyzer.py.
 */

import Database from 'better-sqlite3';

interface GraphInsight {
  type: string;
  title: string;
  description: string;
  node_ids: number[];
  severity: string;
  action: { label: string; endpoint: string; method: string } | null;
}

interface GraphAnalysisResult {
  insights: GraphInsight[];
  stats: { node_count: number; edge_count: number; density: number };
  computed_at: string;
}

export class GraphAnalyzer {
  constructor(private readonly db: Database.Database) {}

  analyze(): GraphAnalysisResult {
    const stats = this.computeStats();
    const insights: GraphInsight[] = [
      ...this.findOrphans(),
      ...this.findHubs(),
      ...this.findClusters(),
      ...this.findStaleNodes(),
    ];
    return { insights, stats, computed_at: new Date().toISOString() };
  }

  private computeStats(): { node_count: number; edge_count: number; density: number } {
    try {
      const nc = (this.db.prepare('SELECT COUNT(*) as c FROM knowledge_entries').get() as { c: number }).c;
      const ec = (this.db.prepare('SELECT COUNT(*) as c FROM knowledge_graph_edges').get() as { c: number }).c;
      const maxEdges = nc > 1 ? nc * (nc - 1) : 1;
      return { node_count: nc, edge_count: ec, density: Math.round((ec / maxEdges) * 10000) / 10000 };
    } catch {
      return { node_count: 0, edge_count: 0, density: 0 };
    }
  }

  private findOrphans(): GraphInsight[] {
    try {
      const rows = this.db.prepare(
        `SELECT e.id FROM knowledge_entries e WHERE e.id NOT IN (
           SELECT source_id FROM knowledge_graph_edges
           UNION SELECT target_id FROM knowledge_graph_edges
         ) LIMIT 20`
      ).all() as { id: number }[];
      if (!rows.length) return [];
      return [{
        type: 'orphans', title: `${rows.length} Orphan Nodes`,
        description: 'Entries không có relationships nào',
        node_ids: rows.map(r => r.id), severity: 'warning',
        action: { label: 'Find Related', endpoint: 'api/kb/entries/{id}/find-related', method: 'POST' },
      }];
    } catch { return []; }
  }

  private findHubs(): GraphInsight[] {
    try {
      const rows = this.db.prepare(
        `SELECT node_id, SUM(cnt) as total FROM (
           SELECT source_id AS node_id, COUNT(*) AS cnt FROM knowledge_graph_edges GROUP BY source_id
           UNION ALL
           SELECT target_id AS node_id, COUNT(*) AS cnt FROM knowledge_graph_edges GROUP BY target_id
         ) GROUP BY node_id HAVING total > 10 ORDER BY total DESC LIMIT 10`
      ).all() as { node_id: number; total: number }[];
      if (!rows.length) return [];
      return [{
        type: 'hubs', title: `${rows.length} Hub Nodes`,
        description: 'Entries có >10 relationships (highly connected)',
        node_ids: rows.map(r => r.node_id), severity: 'info', action: null,
      }];
    } catch { return []; }
  }

  private findClusters(): GraphInsight[] {
    try {
      const edges = this.db.prepare('SELECT source_id, target_id FROM knowledge_graph_edges')
        .all() as { source_id: number; target_id: number }[];
      const nodes = this.db.prepare('SELECT id FROM knowledge_entries')
        .all() as { id: number }[];
      if (!nodes.length) return [];
      const adj = new Map<number, Set<number>>();
      for (const n of nodes) adj.set(n.id, new Set());
      for (const e of edges) {
        adj.get(e.source_id)?.add(e.target_id);
        adj.get(e.target_id)?.add(e.source_id);
      }
      const components = countComponents(adj);
      if (components <= 1) return [];
      return [{
        type: 'clusters', title: `${components} Disconnected Clusters`,
        description: 'Graph có nhiều components tách biệt',
        node_ids: [], severity: 'info', action: null,
      }];
    } catch { return []; }
  }

  private findStaleNodes(): GraphInsight[] {
    const threshold = new Date(Date.now() - 180 * 86400000).toISOString();
    try {
      const rows = this.db.prepare(
        'SELECT id FROM knowledge_entries WHERE updated_at < ? LIMIT 15'
      ).all(threshold) as { id: number }[];
      if (!rows.length) return [];
      return [{
        type: 'stale', title: `${rows.length} Stale Nodes (>180 days)`,
        description: 'Entries chưa được update > 180 ngày',
        node_ids: rows.map(r => r.id), severity: 'warning',
        action: { label: 'Review', endpoint: 'api/kb/entries/{id}/review', method: 'POST' },
      }];
    } catch { return []; }
  }
}

/** Count connected components via BFS. */
function countComponents(adj: Map<number, Set<number>>): number {
  const visited = new Set<number>();
  let components = 0;
  for (const node of adj.keys()) {
    if (visited.has(node)) continue;
    components++;
    const queue = [node];
    while (queue.length) {
      const current = queue.shift()!;
      if (visited.has(current)) continue;
      visited.add(current);
      for (const neighbor of adj.get(current) ?? []) {
        if (!visited.has(neighbor)) queue.push(neighbor);
      }
    }
  }
  return components;
}
